src/pipecat/processors/canonical_metrics_processor.py

```diff
@@ -0,0 +1,202 @@
+        default_part_size (int, optional): Default size for multipart upload parts in bytes. Defaults to 1MB (1024 * 1024 * 1).
+
+    Attributes:
+        call_id (str): Stores the unique call identifier.
+        assistant (str): Stores the assistant identifier.
+        assistant_speaks_first (bool): Indicates whether the assistant speaks first.
+        output_dir (str): Directory path for saving temporary audio files.
+        partsize (int): Size of each part for multipart uploads.
+
+    The constructor also ensures that the output directory exists.
+    This class requires a Canonical API key to be set in the CANONICAL_API_KEY environment variable.
+    """
+
+    def __init__(
+            self,
+            call_id: str,
+            assistant: str,
+            assistant_speaks_first: bool = True,
+            output_dir: str = "recordings",
+            default_part_size: int = 1024 * 1024 * 1):
+        super().__init__()
+        if not os.environ.get("CANONICAL_API_KEY"):
+            raise ValueError(
+                "CANONICAL_API_KEY is not set, a Canonical API key is required to use this class")
+        self.call_id = call_id
+        self.assistant = assistant
+        self.assistant_speaks_first = assistant_speaks_first
+        self.output_dir = output_dir
+        self.partsize = default_part_size
+        self.end_of_call = False
+
+    async def process_frame(self, frame: Frame, direction: FrameDirection):
+        await super().process_frame(frame, direction)
+        if self.end_of_call:
+            return
+
+        if (isinstance(frame, EndFrame) or isinstance(frame, CancelFrame)):
+            self.end_of_call = True
+            if self.has_audio():
+                os.makedirs(self.output_dir, exist_ok=True)
+                filename = self.get_output_filename()
+                with BytesIO() as buffer:
+                    with wave.open(buffer, 'wb') as wf:
+                        wf.setnchannels(self.num_channels)
+                        wf.setsampwidth(self.sample_rate // 8000)
+                        wf.setframerate(self.sample_rate)
+                        wf.writeframes(self.audio_buffer)
+                    wave_data = buffer.getvalue()
+
+                async with aiofiles.open(filename, 'wb') as file:
+                    await file.write(wave_data)
+
+                try:
+                    await self.multipart_upload(filename)
+                    await aiofiles.os.remove(filename)
+                except FileNotFoundError:
+                    pass
+                except Exception as e:
+                    raise e
+                self.audio_buffer = bytearray()
+
+    def get_output_filename(self):
+        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+        return f"{self.output_dir}/{timestamp}-{uuid.uuid4().hex}.wav"
+
+    def canonical_api_url(self):
+        return os.environ.get("CANONICAL_API_URL", "https://voiceapp.canonical.chat/api/v1")
```

Comment:
Same to `api_key`, we could probably add this to the constructor defaulting to `"https://voiceapp.canonical.chat/api/v1"`

Reply:
Done.